src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs

```diff
@@ -0,0 +1,277 @@
+                    {
+                        // return false;
+                        F.CloseMethod(F.Return(F.Literal(false)));
+                        return;
+                    }
+                }
+                else
+                {
+                    MethodSymbol? printMethod = FindValidPrintMethod(ContainingType.BaseTypeNoUseSiteDiagnostics, DeclaringCompilation);
+                    if (printMethod is null)
+                    {
+                        F.CloseMethod(F.ThrowNull()); // an error was reported in base checks already
+                        return;
+                    }
+
+                    var basePrintCall = F.Call(receiver: F.Base(ContainingType.BaseTypeNoUseSiteDiagnostics), printMethod, builder);
+                    if (printableMembers.IsEmpty)
+                    {
+                        // return base.print(builder);
+                        F.CloseMethod(F.Return(basePrintCall));
+                        return;
+                    }
+                    else
+                    {
+                        // if (base.print(builder))
+                        //     builder.Append(", ")
+                        block!.Add(
+                            F.If(basePrintCall,
+                                F.ExpressionStatement(F.Call(receiver: builder, F.WellKnownMethod(WellKnownMember.System_StringBuilder__AppendString), F.StringLiteral(", ")))));
+                    }
+                }
+
+                Debug.Assert(!printableMembers.IsEmpty && block is object);
+
+                for (var i = 0; i < printableMembers.Length; i++)
+                {
+                    // builder.Append(<name>);
+                    // builder.Append(" = ");
+                    // builder.Append((object)<value>);
+                    // builder.Append(", "); // except for last member
+
+                    var member = printableMembers[i];
+                    block.Add(
+                        F.ExpressionStatement(F.Call(receiver: builder,
+                            F.WellKnownMethod(WellKnownMember.System_StringBuilder__AppendString), F.StringLiteral(member.Name))));
+                    block.Add(
+                        F.ExpressionStatement(F.Call(receiver: builder,
+                            F.WellKnownMethod(WellKnownMember.System_StringBuilder__AppendString), F.StringLiteral(" = "))));
+
+                    var value = member.Kind switch
+                    {
+                        SymbolKind.Field => F.Field(F.This(), (FieldSymbol)member),
+                        SymbolKind.Property => F.Property(F.This(), (PropertySymbol)member),
+                        _ => throw ExceptionUtilities.UnexpectedValue(member.Kind)
+                    };
+
+                    block.Add(F.ExpressionStatement(
+                        F.Call(receiver: builder,
+                            F.WellKnownMethod(WellKnownMember.System_StringBuilder__AppendObject),
```

Comment:
I take it we are accepting "records containing reference cycles consisting of only records will crash on ToString()" on the assumption that records containing cycles are very rare/an antipattern?

For a moment I was concerned that a record which has some non-record class property containing a cycle would crash, but that would not happen because the non-record property would be expected to handle cycles in its ToString().

Reply:
> but that would not happen because the non-record property would be expected to handle cycles in its ToString().

Yes.

That said, it's possible to create a cycle (overflowing `ToString`) with a pure record : `record List { public List Next { get; set; } }` (and then set up a cycle). I have a follow-up item to discuss with LDM in the test plan about this.

---
In reply to: [469423674](https://github.com/dotnet/roslyn/pull/46616#discussion_r469423674) [](ancestors = 469423674)

Reply:
Consider ignoring mutable fields and properties. It seems easy to add an explicit `Parent` pointer in particular that would cause problems: `record R { public R Parent; }`.

---
In reply to: [469472601](https://github.com/dotnet/roslyn/pull/46616#discussion_r469472601) [](ancestors = 469472601,469423674)